### PR TITLE
Optimize the implementation of uri & Fix async log bug

### PR DESCRIPTION
**Pull request by you-n-g**
## How Has This Been Tested?

- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [ ] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [ ] Add new feature
- [ ] Update documentation

### CODE REVIEW

**Comment by you-n-g** [comment]
@ChiahungTai  You are welcome to review this PR.

File: qlib/workflow/expm.py [hunk 1]
```diff
@@ -15,23 +15,35 @@
...
 
+
 logger = get_module_logger("workflow")
 
 
 class ExpManager:
     """
-    This is the `ExpManager` class for managing experiments. The API is designed similar to mlflow.
-    (The link: https://mlflow.org/docs/latest/python_api/mlflow.html)
+        This is the `ExpManager` class for managing experiments. The API is designed similar to mlflow.
```
Comment (you-n-g) [inline qlib/workflow/expm.py:24]: @ChiahungTai  You can start from these docs.

File: qlib/workflow/expm.py [hunk 2]
```diff
@@ -15,23 +15,35 @@
...
+        When the user starts an experiment, the user may want to set the uri to a specific uri (it will override **default uri** during this period), and then unset the **specific uri** and fallback to the **default uri**.    `ExpManager._active_exp_uri` is that **specific uri**.
     """
 
+    active_experiment: Optional[Experiment]
+
     def __init__(self, uri: Text, default_exp_name: Optional[Text]):
-        self._current_uri = uri
+        # The `ExpManager` is expected to be a singleton.
+        # So we just use `C.exp_manager["kwargs"]["uri"]` for `uri` instead.
+        # Setting the variable to Config will make the uri consistant for multi-processing in all platforms
```
Comment (you-n-g) [inline qlib/workflow/expm.py:38]:
```suggestion
```

**Comment by ChiahungTai** [comment]
Fail the test in test_all_pipeline.py.
I have add the test case for you in #1363.

File: qlib/workflow/expm.py (hunk 1 above)
Comment (ChiahungTai) [reply qlib/workflow/expm.py:24]: "we can have multiple `Experiment`s with different uri" => IIUC, the words are imply the design of ExpManger is suggestion the user use different uri in each experiment? 

From my understanding in MLFLow, the uri is like a store backend(sql, file...). The ExpManger is a single entry to lookup and manipulate the experiments in the uri.
 
So the user can change different uri to retrieve the different topic of experiements.

**Comment by ChiahungTai** [comment]
Looks like the exception is the same as the test case #1363.

            except MlflowException as e:
                raise ValueError(
                    "No valid experiment has been found, please make sure the input experiment name is correct."
                ) from e

File: qlib/workflow/__init__.py [hunk 3]
```diff
@@ -370,11 +369,11 @@ def uri_context(self, uri: Text):
             the temporal uri
         """
         prev_uri = self.exp_manager.default_uri
-        C.exp_manager["kwargs"]["uri"] = uri
+        self.exp_manager.default_uri = uri
```
Comment (ChiahungTai) [inline qlib/workflow/__init__.py:372]: I think the set of default_uri is not engouh. The self.exp_manager is not change the client uri if you only change default uri.

File: qlib/workflow/expm.py [hunk 4]
```diff
@@ -345,12 +355,11 @@ def start_exp(
...
         return self.active_experiment
 
-    def end_exp(self, recorder_status: Text = Recorder.STATUS_S):
+    def _end_exp(self, recorder_status: Text = Recorder.STATUS_S):
         if self.active_experiment is not None:
             self.active_experiment.end(recorder_status)
             self.active_experiment = None
-        # When an experiment end, we will release the current uri.
-        self._current_uri = None
+        self._set_client_uri()
```
Comment (ChiahungTai) [inline qlib/workflow/expm.py:362]: Pump into an exception - No argument.

File: qlib/workflow/expm.py [hunk 5]
```diff
@@ -322,18 +333,17 @@ def client(self):
             self._client = mlflow.tracking.MlflowClient(tracking_uri=self.uri)
         return self._client
```
Comment (ChiahungTai) [inline qlib/workflow/expm.py:334]: I think you might need to sync the returned self._client to the updated default_uri. 
Or you have set the uri in every function in access the client.

File: qlib/workflow/expm.py (hunk 1 above)
Comment (you-n-g) [reply qlib/workflow/expm.py:24]: Yes.  
ExpManager is a singleton, but we can change its uri  to get experiments from different uri.
I added some comments just now.
Please check if it is helpful

File: qlib/workflow/recorder.py [hunk 6]
```diff
@@ -329,7 +329,7 @@ def get_local_dir(self):
     def start_run(self):
         # set the tracking uri
         mlflow.set_tracking_uri(self.uri)
-        # start the run
+        # start the RuntimeError
```
Comment (ChiahungTai) [inline qlib/workflow/recorder.py:332]: What does the start the RuntimeError mean?

File: tests/dependency_tests/test_mlflow.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        we don't cache _client (this is helpful to reduce maintainance work when MLflowExpManager's uri is chagned)
+
+        This implementation is based on the assumption creating a client is fast
+        """
+        start = time.time()
+        for i in range(10):
+            _ = mlflow.tracking.MlflowClient(tracking_uri=str(self.TMP_PATH))
+        end = time.time()
+        elasped = end - start
+        self.assertGreater(1e-2, elasped)  # it can be done in less than 10ms
```
Comment (ChiahungTai) [inline tests/dependency_tests/test_mlflow.py:29]: assertGreater or assertLesser

File: qlib/workflow/expm.py (hunk 1 above)
Comment (ChiahungTai) [reply qlib/workflow/expm.py:24]: Your new comment is more clear than before.

File: qlib/workflow/__init__.py (hunk 3 above)
Comment (you-n-g) [reply qlib/workflow/__init__.py:372]: It is fixed now.
The _client will not be cached.  No further maintenance is required for it.

File: qlib/workflow/expm.py (hunk 5 above)
Comment (you-n-g) [reply qlib/workflow/expm.py]: Fixed Now.

File: tests/dependency_tests/test_mlflow.py (hunk 7 above)
Comment (you-n-g) [reply tests/dependency_tests/test_mlflow.py]: I have changed the direction. And it is more readable than before

**Comment by you-n-g** [comment]
@ChiahungTai 
Hi, I have fixed all the comments and waiting for the CI.

I checked out [your test](https://github.com/microsoft/qlib/pull/1363/files#diff-e85369ca90d47fb4376b973050e0f2bd66cca8aa16ab93949415c3af58b2a288R85)

I didn't get what you want to assert in the test.

It seems that the exception looks reasonable.
When we want to get a recorder,  Qlib tries to get an experiment first.   The experiment name or id is not given.
So the default exp name is used. 
There is no experiment which is active or is named with the default exp name.
So an exception of `No valid experiment has been found` is raised

File: qlib/workflow/recorder.py (hunk 6 above)
Comment (you-n-g) [reply qlib/workflow/recorder.py]: Fixed now

File: qlib/workflow/expm.py (hunk 4 above)
Comment (you-n-g) [reply qlib/workflow/expm.py]: Fixed now.

**Comment by ChiahungTai** [comment]
The test code is pass after assign the experiment_name and recorder_name.
But in my own project, I provide the experiment_name and recorder_name.
But the code is too long to write a simple test case.
`
    with R.uri_context(uri=uri_path):
        recorder = R.get_recorder(experiment_name=EXP_NAME, recorder_name="original")
`
I am working on another version of cached_pipeline in my own project. I might test the new commit and try it my local first.
Also I will upload a new test if I can rewrite a simple test case.

> @ChiahungTai Hi, I have fixed all the comments and waiting for the CI.
> 
> I checked out [your test](https://github.com/microsoft/qlib/pull/1363/files#diff-e85369ca90d47fb4376b973050e0f2bd66cca8aa16ab93949415c3af58b2a288R85)
> 
> I didn't get what you want to assert in the test.
> 
> It seems that the exception looks reasonable. When we want to get a recorder, Qlib tries to get an experiment first. The experiment name or id is not given. So the default exp name is used. There is no experiment which is active or is named with the default exp name. So an exception of `No valid experiment has been found` is raised

**Comment by ChiahungTai** [comment]
Your new patch works fine in my own project.

> The test code is pass after assign the experiment_name and recorder_name. But in my own project, I provide the experiment_name and recorder_name. But the code is too long to write a simple test case. `with R.uri_context(uri=uri_path): recorder = R.get_recorder(experiment_name=EXP_NAME, recorder_name="original")` I am working on another version of cached_pipeline in my own project. I might test the new commit and try it my local first. Also I will upload a new test if I can rewrite a simple test case.
> 
> > @ChiahungTai Hi, I have fixed all the comments and waiting for the CI.
> > I checked out [your test](https://github.com/microsoft/qlib/pull/1363/files#diff-e85369ca90d47fb4376b973050e0f2bd66cca8aa16ab93949415c3af58b2a288R85)
> > I didn't get what you want to assert in the test.
> > It seems that the exception looks reasonable. When we want to get a recorder, Qlib tries to get an experiment first. The experiment name or id is not given. So the default exp name is used. There is no experiment which is active or is named with the default exp name. So an exception of `No valid experiment has been found` is raised

**Comment by ChiahungTai** [comment]
The patch LGTM. You can merge it after fix the pylint error.

**Pull request merged**